### PR TITLE
feat: expose @sap-ux/jest-environment-ui5

### PR DESCRIPTION
**Pull request by nlunets**
The module is intended to provide a way to test ui5 applications using jest.

This works by creating a custom test environment in which ui5 will be loaded and the test will be run.

The code existed internally before now and this will be notified as such in the Changelog.

### CODE REVIEW

File: packages/jest-environment-ui5/src/shim/hasher.js [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+					var frameDoc = _frame.contentWindow.document;
+					frameDoc.open();
+					//update iframe content to force new history record.
+					//based on Really Simple History, SWFAddress and YUI.history.
+					frameDoc.write(
+						"<html><head><title>" +
+							document.title +
+							'</title><script type="text/javascript">var frameHash="' +
+							_hash +
+							'";</script></head><body>&nbsp;</body></html>'
```
[the remark this answers is not in the document]
Comment (nlunets) [reply packages/jest-environment-ui5/src/shim/hasher.js:105]: This file is just a shim of the ui5 code to patch some behavior of the hasher lib that behaves differently when it detects a node environment

File: packages/jest-environment-ui5/src/shim/hasher.js [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+				return path;
+			}
+
+			function _encodePath(path) {
+				//used encodeURI instead of encodeURIComponent to preserve '?', '/',
+				//'#'. Fixes Safari bug [issue #8]
+				path = encodeURI(path);
+				if (_isIE && _isLocal) {
+					//fix IE8 local file bug [issue #6]
+					path = path.replace(/\?/, "%3F");
```
[the remark this answers is not in the document]
Comment (nlunets) [reply packages/jest-environment-ui5/src/shim/hasher.js:202]: This file is just a shim of the ui5 code to patch some behavior of the hasher lib that behaves differently when it detects a node environment

**Review by tobiasqueck — CHANGES_REQUESTED** [review]
* why do we need the shims, and is it ok to add the code like that to the repo
* please add a short descriptive comment to each function
* please check your package.json, you refer to the `dist` folder as content (not to `src`) but you don't have a build script (or anything else) that would create `dist`
* please check the lint warnings, all of them are fixable

**Comment by nlunets** [comment]
Hello @tobiasqueck  
- why do we need the shims, and is it ok to add the code like that to the repo
  -> I found a way to get rid of them 
- please add a short descriptive comment to each function
   -> done
please check your package.json, you refer to the dist folder as content (not to src) but you don't have a build script (or anything else) that would create dist
   -> done
please check the lint warnings, all of them are fixable
   -> kept one that didn't make sense to fix in my opinion

**Comment by tobiasqueck** [comment]
> Hello @tobiasqueck
> 
> * why do we need the shims, and is it ok to add the code like that to the repo
>   -> I found a way to get rid of them
> * please add a short descriptive comment to each function
>   -> done
>   please check your package.json, you refer to the dist folder as content (not to src) but you don't have a build script (or anything else) that would create dist
>   -> done
>   please check the lint warnings, all of them are fixable
>   -> kept one that didn't make sense to fix in my opinion

Thanks! The code is easier to understand now. Please have a look at the following issues:
* I see one file left in shims, could you please enhance the `/.reuse` file with information where it is from and what you did as well as under which license it is
* could you please use the folders fixtures and unit (in test) for your sample and your tests 
* seems like prettier didn't run on your jest tests
* Shouldn't `jest` be a devDependency since you use it in one of your tests with `const jestCLI = require('jest');`

**Comment by nlunets** [comment]
@tobiasqueck all done i think :)

**Review by tobiasqueck — APPROVED** [review]
* changeset good
* general project structure adheres to the rest of the repo
* code is understandable - sadly in JS instead of TS but it is ok for now, since this is an existing module
* code is complex but still understandable and maintainable

**Comment by tobiasqueck** [comment]
@Jimmy-Joseph19 please try if the jest-environment works for you as expected.

**Review by tobiasqueck — APPROVED** [review]

**Review by tobiasqueck — DISMISSED** [review]
Re-approving the PR. All changes since the last review improve the code quality and don't affect the functionality.

**Review by tobiasqueck — APPROVED** [review]

**Review by voicis — APPROVED** [review]
@nlunets thanks!
- Tested with a basic scenario in `preview-middleware-client`
- Documentation and error handling could be improved, but that is something we can work in the future

**Pull request merged**